### ISSUE TITLE
MisMatch of Vertices for Surface-Matched Surfaces

### ISSUE THREAD

**Issue by Ski90Moo** [openstudiocoalition/openstudio-sketchup-plugin#60]
This vertex mismatch error is very common where matched surfaces are comprised of more that four vertices:

** Severe  ** RoofCeiling:Detailed="SURFACE ???", Vertex size mismatch between base surface :SURFACE ??? and outside boundary surface: SURFACE ???

I suspect Sketchup defines a surface not just based on its component vertices, but also shared vertices.  So, for example, if two surfaces have a shared vertex, only one of the surfaces will have that vertex in its definition.  The other surface will be missing the vertex in it's definition.

Here is a detailed video of the problem and work-around I developed to resolve the issue: https://1drv.ms/v/s!AseQmgZxffwzgbN01VYhYjiIjfhXaQ?e=62GZTC

**Comment by Mkellyeng** [comment]
I've run into this issue several times, as well

I'm not quite sure what causes the issue, but I have an extra detail to add
Using the example from your video:
You're wanting the following surface, in Red (11 vertices)
![image](https://user-images.githubusercontent.com/46730121/113961636-b88a6e00-9869-11eb-83be-871fcb2e13e5.png)

But I think the .osm/.idf is actually using the following surface, in Green (7 vertices)
![image](https://user-images.githubusercontent.com/46730121/113961711-d0fa8880-9869-11eb-8a8f-80c967d45638.png)

That is, it drops the 4 vertices from the middle rectangle, leaving you with an incorrect area, and an incorrect number of vertices, when compared to its matched surface. 

Note that this is tricky to spot visually, because it's seemingly _behind_ the excluded rectangle. 

I believe your workaround corrects the issue because you used one of the vertices that was being lost. For whatever reason, that seems to help them get found?
Eg: If you subdivided the surface by drawing the Yellow line, the issue would not be resolved. But if you drew the Pink line, it would be
![image](https://user-images.githubusercontent.com/46730121/113962713-9265cd80-986b-11eb-945c-6971ecd0a93a.png)

**Comment by Ski90Moo** [comment]
@Mkellyeng, I agree.  OpenStudio is just importing the vertices from SketchUp.  In the example above, SketchUp is telling OS that the green surface only has 7 vertices and the red surface has 4 vertices.  I have a theory:

SketchUp is doing some sort of geometric optimization to save computing power.  In other words, it is saying; why define a surface with all of its constituent vertices if it shares vertices with a co-planer surface?  Dropping those four shared vertices from one of the co-planer surfaces reduces the amount of extraneous data.

If you copy and paste the green surface to somewhere else in SketchUp, it still shows 11 vertices...there seems to be something lost in translation when going from SketchUp to OpenStudio.

**Comment by macumber** [comment]
Thanks for the detailed report.  SketchUp has Faces which have an outer loop and then potentially inner loops.  We detect these inner loops and classify them as doors or windows on walls.  I think there is an issue here because you have basically drawn a door but on a horizontal floor surface so we aren't classifying it as a door.  Plus you want it to be a separate surface and not a subsurface.  Dividing the surface so there are three faces with no inner loops would work.  I think the only thing we could try to do in the plug-in would be to detect the case of a door on horizontal floor surface and break up the outer loop. 

PS: nice SketchUp selection skills!

**Comment by Mkellyeng** [comment]
Interesting, how do inner loops work exactly?

Something like this?
![image](https://user-images.githubusercontent.com/46730121/114830685-d3338880-9e0f-11eb-9e11-853677ab647a.png)

Do they need to share an edge with the Outer Loop? 

Now that I think about it, I think I've usually seen this vertex-mismatch issue on the _edges of surfaces_, perhaps that edge is not being subdivided properly?

**Comment by Ski90Moo** [comment]
Thanks guys, this would make sense.  I wonder if this is related to the issue with sky lights?  Sometimes the ceilings are miss-identified as skylights.

**Comment by WangDAL-stack** [comment]
I got the same questions in this situation....When I was trying to intersect and match those surfaces,not every piece of surface turn into green,but green and blue(under the circumstance of Render By Boundary Bondition).
These problems stuck me a long period so that I can not to finish other steps of os.
![图片](https://user-images.githubusercontent.com/98703530/151708899-351a10ae-ec80-4b5d-8c15-b9eede77bcf5.png)
Here is the predict of the problems and submitting my sincere appreciation beforehand.

**Comment by Ski90Moo** [comment]
> I got the same questions in this situation....When I was trying to intersect and match those surfaces,not every piece of surface turn into green,but green and blue(under the circumstance of Render By Boundary Bondition). These problems stuck me a long period so that I can not to finish other steps of os. ![图片] Here is the predict of the problems and submitting my sincere appreciation beforehand.

This looks like your polygons only have four vertices, so I don't think this issue would apply to your problem.  Try solving like this: https://youtu.be/FfIpsCsO7HM

**Comment by Mkellyeng** [comment]
@WangDAL-stack Yeah, your issue here is a little different I think. The original issue is to do with surfaces not getting **intersected and subdivided** properly

 I think your issue is to do with surfaces not **matching** properly, which is something different. I'm not sure why that occasionally fails to match properly, but if you need a simple, manual fix, you can use the tips in the link that @Ski90Moo posted. You can use the inspection tool to manually set the outside boundary condition. Alternatively, you can set it using a text editor, which is very easy

**Comment by Mkellyeng** [comment]
@WangDAL-stack Can you share your file? It looks like you have a nice, simple example of that surface matching issue

**Comment by ianvg** [comment]
Interesting read, I guess there's no way that Sketchup can be 'asked' to not do that geometric optimization, whereby vertices are lost. I was going to post a new issue ([related to this Unmet Hours Forum post about subsurfaces](https://unmethours.com/question/3409/subsurface-is-missing-its-base-surface-it-cannot-be-drawn/)), but the problem described here I _think_ encompass it? I am personally having trouble creating inner loops that share an edge with the parent surface. Is there any related issues or programming that will eventually address this issue?

If not, I suppose the solution is to slightly offset doors and windows from the parent edges to create proper subsurfaces.

**Comment by macumber** [comment]
The OpenStudio SketchUp Plug-In uses SketchUp's assignment of loops to faces to decide if a face is a surface or a sub surface.  If two faces share an edge then SketchUp assigns separate loops to each face.  I don't know of any way to override this behavior.  The recommended solution would be to offset the sub surfaces from the surface edges by at least an inch on all sides.
